doc/adr/0004-roll-our-own-kubernetes-profile.md

```diff
@@ -0,0 +1,53 @@
+# 4. Roll our own kubernetes profiles
+
+Date: 2019-07-25
+
+Status
+------
+
+Accepted
+
+Context
+-------
+
+Puppetlabs maintains [a kubernetes module][1] in the forge, and Rancher
+maintains software for managing multiple clusters running anywhere.
+
+The puppetlabs module claims to support Debian, but, in practice, it
+appears to only support Ubuntu. We tried forking it and making changes
+to get it working, but it was a rabbithole of error after error. Also,
+their solution to managing SSL keys was to generate them locally and
+store them all in hiera, which was cumbersome. They also didn't support
+managing the CIDRs for the internal network.
+
+Rancher seemed very nice at first, but it was unstable in practice.
+Despite running a highly available set of control nodes, bringing one
+down did in fact break the cluster. Also, if /var/lib/docker filled up
+on any machine, the only solution was to destroy and recreate the entire
+cluster.
+
+My assessment of the puppetlabs module is that it tries to do too much
+on its own. My assessment of rancher is that it's not yet stable enough
+to rely on.
+
+[1]: https://forge.puppet.com/puppetlabs/kubernetes
+
+Decision
+--------
+
+We will roll our own kubernetes profiles instead of relying on someone
+else's solution.
+
+Consequences
+------------
+
+While our profiles are sufficient for creating an environment where
+kubernetes can flourish, bootstrapping must be done outside of puppet,
```

Comment:
If we spin up new kubernetes nodes and set up the hieradata so that puppet sets them up, then, after an hour or so, they'll be ready to be a cluster (the right ports will be open to the right machines, all the needed configuration will be in the right place), but kubernetes itself won't be running. There are a couple commands that one of us will have to run on each machine in the cluster and in a particular order, and we'll have to go through that once per cluster.

So to answer your question, "bootstrapping" is going from "puppet's done configuring these machines to be kubernetes" to "kubernetes is running on these machines." It needs to happen once every time a new cluster is created, and I'm currently punting on whether/how that process should be automated.

It could in theory be automated by puppet (it's what their module does), but I'd rather avoid doing that here, mostly because it feels like it'd be a lot of work (getting puppet to pick one machine to run one command on, look at the output, and then run other commands on other machines based on that output, eeurggh, _and_ never run them ever again on those machines) just to avoid a little repetitive work on our part.

Reply:
I should add that the bootstrapping process isn't documented (in part because I'm punting on it), but we do have [a ticket for that in jira](https://tools.lib.umich.edu/jira/browse/AEIM-2176) that at least acknowledges that I haven't done this yet.

Reply:
Thanks for the detail -- all sounds good.